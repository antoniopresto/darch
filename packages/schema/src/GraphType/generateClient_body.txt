
export type GraphqlClientHelpers = typeof graphqlClientHelpers;
export type GraphQLEntry = GraphqlClientHelpers[keyof GraphqlClientHelpers];

export type GraphQLFetchParams<K extends GraphQLEntry['name']> = {
  operationInfo: GraphqlClientHelpers[K];
  operationName: K;
  getBody: (args: ExpectedGraphQLClient[K]['args']) => {
    query: string;
    variables: Record<string, any>;
    operationName: K;
  };
  parseArgs: (args: ExpectedGraphQLClient[K]['args']) => Record<string, any>;
  mountBodyString(args: ExpectedGraphQLClient[K]['args']): string;
};

export function getGraphQLFetchHelpers<MethodName extends GraphQLEntry['name']>(
  methodName: MethodName,
): GraphQLFetchParams<MethodName> {
  const helpers = graphqlClientHelpers[methodName];

  function parseArgs(args: any) {
    const vars: Record<string, any> = {};
    const parsedArgs: any = helpers.input.parse(args || {}, (_, error) => {
      return `\nGraphQLClientArgumentsError: method ${methodName}: \n${error.message}`;
    });

    Object.entries(helpers.operation.varNames).forEach(([inputVarName, { varName }]) => {
      vars[varName] = parsedArgs[inputVarName];
    });

    return vars;
  }

  function getBody(args: any) {
    const variables = parseArgs(args);
    return {
      query: helpers.operation.query,
      variables,
      operationName: methodName,
    };
  }

  function mountBodyString(args: any) {
    const body = getBody(args);
    return JSON.stringify(body);
  }

  return {
    operationInfo: helpers,
    operationName: methodName,
    parseArgs,
    getBody,
    mountBodyString,
  };
}
