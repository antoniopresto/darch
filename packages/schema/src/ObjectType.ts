import { RuntimeError } from '@darch/utils/lib/RuntimeError';
import { StrictMap } from '@darch/utils/lib/StrictMap';
import { dynamicRequire } from '@darch/utils/lib/dynamicRequire';
import { isProduction } from '@darch/utils/lib/env';
import { expectedType } from '@darch/utils/lib/expectedType';
import { getTypeName } from '@darch/utils/lib/getTypeName';
import { invariantType } from '@darch/utils/lib/invariant';
import { isBrowser } from '@darch/utils/lib/isBrowser';
import { simpleObjectClone } from '@darch/utils/lib/simpleObjectClone';
import { ForceString } from '@darch/utils/lib/typeUtils';
import type {
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLObjectType,
} from 'graphql';

import type { DarchType } from './DarchType';
import type { ParseTypeOptions } from './GraphQLParser/GraphQLParser';
import type { ParseInterfaceOptions } from './GraphQLParser/GraphQLParser';
import type { Infer } from './Infer';
import type { ObjectDefinitionInput } from './TObjectConfig';
import {
  parseValidationError,
  ValidationCustomMessage,
} from './applyValidator';
import { assertSameDefinition } from './assertSameDefinition';
import {
  getObjectDefinitionMetaField,
  isMetaFieldKey,
  MetaFieldDef,
  objectMetaFieldKey,
  Serializable,
} from './fields/MetaFieldField';
import type {
  FinalObjectDefinition,
  ObjectFieldInput,
  ParseFields,
  ToFinalField,
} from './fields/_parseFields';
import { validateObjectFields } from './getObjectErrors';
import { getObjectHelpers, ObjectHelpers } from './getObjectHelpers';
import type { ObjectToTypescriptOptions } from './objectToTypescript';
import { parseObjectDefinition } from './parseObjectDefinition';
import { withCache, WithCache } from './withCache';

export { RuntimeError } from '@darch/utils/lib/RuntimeError';
export * from './parseObjectDefinition';
export * from './objectInferenceUtils';
export * from './implementObject';

export class ObjectType<DefinitionInput extends ObjectDefinitionInput> {
  private readonly __definition: any;

  __withCache: WithCache<{
    graphqlInputType: GraphQLInputObjectType;
    graphqlType: GraphQLObjectType;
    graphqlInterfaceType: GraphQLInterfaceType;
    helpers: ObjectHelpers<DefinitionInput>;
  }>;

  constructor(objectDef: DefinitionInput) {
    const parsed = parseObjectDefinition(objectDef);
    this.__definition = parsed.definition;
    this.__withCache = withCache(this);
  }

  get definition(): ParseFields<DefinitionInput> {
    return this.__definition;
  }

  get description() {
    return this.meta.description;
  }

  get meta(): MetaFieldDef {
    return this.__definition[objectMetaFieldKey].def;
  }

  __setMetaData(k: keyof MetaFieldDef, value: Serializable) {
    this.__definition[objectMetaFieldKey].def[k] = value;
  }

  parse(
    input: any,
    options?: {
      customMessage?: ValidationCustomMessage;
    }
  ): Infer<DefinitionInput>;

  parse(
    input: any,
    options?: {
      partial: true;
      customMessage?: ValidationCustomMessage;
    }
  ): Partial<Infer<DefinitionInput>>;

  parse<Fields extends (keyof DefinitionInput)[]>(
    input: any,
    options: {
      customMessage?: ValidationCustomMessage;
      fields: Fields;
    }
  ): {
    [K in keyof Infer<DefinitionInput> as K extends Fields[number]
      ? K
      : never]: Infer<DefinitionInput>[K];
  };

  parse(input: any, options?: any) {
    const { customMessage } = options || {};
    const { errors, parsed } = this.safeParse(input, options);

    if (errors.length) {
      const err: any = parseValidationError(
        input,
        customMessage,
        errors.join(' \n')
      );
      err.isObjectValidationError = true;
      err.fieldErrors = errors;
      throw err;
    }

    return parsed as any;
  }

  validate(input: any): input is Infer<DefinitionInput> {
    try {
      this.parse(input);
      return true;
    } catch (e) {
      return false;
    }
  }

  safeParse(
    input: any,
    options?: {
      partial?: boolean;
      customMessage?: ValidationCustomMessage;
      fields?: keyof DefinitionInput[];
    }
  ): { errors: string[]; parsed: unknown } {
    const { partial = false, fields = Object.keys(this.definition) } =
      options || {};

    const ObjectConstructor: any = this.constructor;

    const errors: string[] = [];
    const parsed: any = {};

    if (!input || typeof input !== 'object' || Array.isArray(input)) {
      throw new RuntimeError(
        `Invalid input. Expected object, found ${getTypeName(input)}`,
        {
          input,
        }
      );
    }

    (fields as string[]).forEach((currField) => {
      if (isMetaFieldKey(currField)) return;

      const value = input[currField];

      const fieldDef = this.definition[currField];

      if (value === undefined && partial) {
        return;
      }

      const result = validateObjectFields({
        createObjectType: (def) => new ObjectConstructor(def),
        fieldName: currField,
        definition: fieldDef,
        value,
      });

      if (Object.prototype.hasOwnProperty.call(input, currField)) {
        parsed[currField] = result.parsed;
      }

      errors.push(...result.errors);
    });

    return { parsed, errors };
  }

  describe(
    ...descriptions:
      | [comment: string]
      | [{ [K in keyof DefinitionInput]?: string }]
  ): ObjectType<DefinitionInput> {
    if (descriptions.length === 1 && typeof descriptions[0] === 'string') {
      this.__setMetaData('description', descriptions[0]);
      return this;
    }

    const commentsConfig = descriptions[0];

    invariantType({ commentsConfig }, 'object', { commentsConfig });

    const definition: FinalObjectDefinition = this.definition as any;

    Object.entries(commentsConfig).forEach(([name, comment]) => {
      invariantType(
        { [name]: definition[name] },
        'object',
        `"${name}" is not in object definition.`
      );
      definition[name].description = comment || '';
    });

    return this;
  }

  removeField<K extends ForceString<keyof DefinitionInput>>(
    field: K | K[]
  ): OmitDefinitionFields<DefinitionInput, K> {
    const fields: string[] = Array.isArray(field) ? field : [field];
    const clone = this.clone();

    for (const k in clone.__definition) {
      if (fields.includes(k)) {
        delete clone.__definition[k];
      }
    }

    return clone as any;
  }

  addFields<T extends ObjectDefinitionInput>(
    definition: T
  ): ObjectExtendDefinition<ParseFields<DefinitionInput>, T> {
    return this.clone(definition) as any;
  }

  makeOptional<K extends ForceString<keyof DefinitionInput>>(
    fields: K | K[]
  ): ObjectType<MakeOptional<DefinitionInput, K>> {
    const fieldList = Array.isArray(fields) ? fields : [fields];
    const clone = this.clone();
    fieldList.forEach((key) => (clone.__definition[key].optional = true));
    return clone as any;
  }

  makeRequired<K extends ForceString<keyof DefinitionInput>>(
    fields: K | K[]
  ): ObjectType<MakeRequired<DefinitionInput, K>> {
    const fieldList = Array.isArray(fields) ? fields : [fields];
    const clone = this.clone();
    fieldList.forEach((key) => (clone.__definition[key].optional = false));
    return clone as any;
  }

  get __isDarchObject(): true {
    return true;
  }

  clone(name?: string): this;

  clone<T extends ObjectDefinitionInput>(
    extend: T,
    name?: string
  ): ObjectExtendDefinition<ParseFields<DefinitionInput>, T>;

  clone<T extends Record<string, ObjectFieldInput | null>>(
    extend: (current: ParseFields<DefinitionInput>) => T,
    name?: string
  ): ObjectType<{ [K in keyof ExcludeNull<T>]: ExcludeNull<T>[K] }>;

  clone<K extends keyof DefinitionInput>(
    fields: K[],
    name?: string
  ): ObjectType<CloneFields<ParseFields<DefinitionInput>, K>>;

  clone<
    K extends keyof DefinitionInput,
    T extends Record<string, ObjectFieldInput | null>
  >(
    fields: K[],
    extend: (current: CloneFields<ParseFields<DefinitionInput>, K>) => T,
    name?: string
  ): ObjectType<{ [K in keyof ExcludeNull<T>]: ExcludeNull<T>[K] }>;

  clone(...args: any[]) {
    const lastArg = args[args.length - 1];

    let newId;
    if (typeof lastArg === 'string') {
      newId = args.pop();
    }

    const definitionClone = simpleObjectClone(this.definition);
    delete definitionClone[objectMetaFieldKey];

    let extend;
    let fields;

    if (args.length === 1) {
      if (Array.isArray(args[0])) {
        fields = args[0];
      } else {
        extend = args[0];
      }
    }

    if (args.length === 2) {
      fields = args[0];
      extend = args[1];
    }

    if (Array.isArray(fields)) {
      for (const k in definitionClone) {
        if (!fields.includes(k)) {
          delete definitionClone[k];
        }
      }
    }

    const def =
      typeof extend === 'function'
        ? extend(definitionClone)
        : typeof extend === 'object'
        ? { ...definitionClone, ...extend }
        : definitionClone;

    Object.keys(def).forEach((k) => {
      if (!def[k]) {
        delete def[k];
      }
    });

    const object = createObjectType(def);

    if (newId) {
      object.identify(newId);
    }

    return object as any;
  }

  get id() {
    return this.meta.id;
  }

  get nonNullId() {
    const id = this.meta.id!;
    if (!id) {
      throw new RuntimeError('Expected object to be identified.', {
        definition: this.definition,
      });
    }
    return id;
  }

  identify<ID extends string>(id: ID): this & { id: ID } {
    if (id && id === this.id) return this as any;

    if (this.id) {
      throw new Error(
        `Trying to replace existing id "${this.id}" with "${id}". You can clone it to create a new Object.`
      );
    }

    expectedType({ id }, 'string', 'truthy');

    if (ObjectType.register.has(id) && ObjectType.register.get(id) !== this) {
      console.error(`Object with id "${id}" already registered.`);
    }

    this.__setMetaData('id', id);
    ObjectType.register.set(id, this);

    return this as any;
  }

  helpers = () => {
    return this.__withCache('helpers', () => getObjectHelpers(this));
  };

  toGraphQL = (name?: string) => {
    if (isBrowser() || typeof module?.require !== 'function') {
      throw new Error('GraphQL transformation is not available in browser.');
    }

    if (name) {
      this.identify(name);
    }

    if (!this.id) {
      throw new RuntimeError(
        'Should object.identify() before converting to Graphql.' +
          '\nYou can call object.clone() to choose a different identification.',
        { 'used definition': this.definition }
      );
    }

    const { GraphQLParser } = ObjectType.serverUtils().graphqlParser;

    return GraphQLParser.objectToGraphQL({
      object: this,
    });
  };

  graphqlType = (options?: ParseTypeOptions): GraphQLObjectType => {
    return this.__withCache('graphqlType', () =>
      this.toGraphQL().getType(options)
    );
  };

  graphqlInterfaceType = (
    options?: ParseInterfaceOptions
  ): GraphQLInterfaceType => {
    return this.__withCache('graphqlInterfaceType', () =>
      this.toGraphQL().interfaceType(options)
    );
  };

  graphqlPrint = (): string => {
    return this.toGraphQL().typeToString();
  };

  typescriptPrint = (options?: ObjectToTypescriptOptions): Promise<string> => {
    return ObjectType.serverUtils().objectToTypescript.objectToTypescript(
      this.nonNullId,
      this,
      options
    );
  };

  graphqlTypeToString = (): string => {
    return this.toGraphQL().typeToString();
  };

  graphqlInputType = (options?: ParseTypeOptions) => {
    return this.__withCache('graphqlInputType', () =>
      this.toGraphQL().getInputType(options)
    );
  };

  static serverUtils() {
    if (isBrowser() || typeof module?.require !== 'function') {
      throw new Error('Server utils are not available on browser.');
    }

    return {
      graphql: dynamicRequire('graphql', module) as typeof import('graphql'),

      graphqlParser: dynamicRequire(
        './GraphQLParser/GraphQLParser',
        module
      ) as typeof import('./GraphQLParser/GraphQLParser'),

      DarchType: dynamicRequire(
        './DarchType',
        module
      ) as typeof import('./DarchType'),

      objectToTypescript: dynamicRequire(
        './objectToTypescript',
        module
      ) as typeof import('./objectToTypescript'),
    };
  }

  static async reset() {
    if (typeof window === 'undefined') {
      const { graphqlParser, DarchType } = ObjectType.serverUtils();
      graphqlParser.GraphQLParser.reset();
      DarchType.DarchType.reset();
    }

    ObjectType.register.clear();
  }

  static register = new StrictMap<string, any>();

  /**
   * Get an Object with the provided id
   *    or set a new Object in the register if not found.
   * @param id
   * @param def
   */
  static getOrSet = <T extends ObjectDefinitionInput>(
    id: string,
    def: T
  ): ObjectType<T> => {
    const existing =
      ObjectType.register.has(id) &&
      (ObjectType.register.get(id) as ObjectType<T>);

    if (existing) {
      !isProduction() && assertSameDefinition(id, def, existing.definition);
      return existing;
    }

    return new ObjectType<T>(def).identify(id);
  };

  static createType<Definition extends ObjectFieldInput>(
    definition: Definition
  ): DarchType<Definition>;

  static createType<Definition extends ObjectFieldInput>(
    name: string,
    definition: Definition
  ): DarchType<Definition>;

  static createType(...args: any) {
    return ObjectType.serverUtils().DarchType.createType(
      // @ts-ignore
      ...args
    );
  }
}

export const DarchObject = ObjectType;

export function createObjectType<
  DefinitionInput extends Readonly<ObjectDefinitionInput>
>(fields: Readonly<DefinitionInput>): ObjectType<DefinitionInput>;

export function createObjectType<
  DefinitionInput extends Readonly<ObjectDefinitionInput>
>(name: string, fields: DefinitionInput): ObjectType<DefinitionInput>;

export function createObjectType<
  DefinitionInput extends Readonly<ObjectDefinitionInput>
>(
  ...args: [string, DefinitionInput] | [DefinitionInput]
): ObjectType<DefinitionInput> {
  const fields = args.length === 2 ? args[1] : args[0];

  const id = args.length === 2 ? args[0] : undefined;
  if (id) return ObjectType.getOrSet(id, fields);

  const idFromDefinition = getObjectDefinitionMetaField(fields)?.def?.id;
  if (idFromDefinition) return ObjectType.getOrSet(idFromDefinition, fields);

  return new ObjectType<DefinitionInput>(fields);
}

export { createObjectType as createDarchObject };

type OmitDefinitionFields<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? ObjectType<{ [K in keyof T as K extends Keys ? never : K]: T[K] }>
  : never;

type ObjectExtendDefinition<T, Ext> = T extends { [K: string]: any }
  ? Ext extends { [K: string]: any }
    ? ObjectType<{
        [K in keyof (T & Ext)]: (T & Ext)[K];
      }>
    : never
  : never;

type CloneFields<T, Keys> = T extends {
  [K: string]: any;
}
  ? { [K in keyof T as K extends Keys ? K : never]: T[K] }
  : never;

type MakeOptional<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? {
      [K in keyof T]: K extends Keys
        ? ToFinalField<T[K]> extends infer Obj
          ? { [K in keyof Obj]: K extends 'optional' ? true : Obj[K] }
          : never
        : T[K];
    }
  : never;

type MakeRequired<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? {
      [K in keyof T]: K extends Keys
        ? ToFinalField<T[K]> extends infer Obj
          ? { [K in keyof Obj]: K extends 'optional' ? false : Obj[K] }
          : never
        : T[K];
    }
  : never;

type ExcludeNull<T> = {
  [K in keyof T as [T[K]] extends [null]
    ? never
    : [T[K]] extends [undefined]
    ? never
    : K]: Exclude<T[K], null>;
};
