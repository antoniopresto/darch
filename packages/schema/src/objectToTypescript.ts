import { jsonToTypescript } from '@darch/utils/lib/jsonToTypescript';

import { ObjectDefinitionInput } from './TObjectConfig';
import { ObjectLike } from './fields/IObjectLike';
import { objectToJSON } from './objectToJSON';

export type ObjectToTypescriptOptions = {
  bannerComment?: string;
  format?: boolean;
  unreachableDefinitions?: boolean;
  additionalProperties?: boolean;
  strictIndexSignatures?: boolean;
};

const defaultBannerComment = ``;

/**
 * Creates a typescript declaration from an object
 * @param name
 * @param object
 * @param options
 */
export async function objectToTypescript(
  name: string,
  object: ObjectLike | ObjectDefinitionInput,
  options?: ObjectToTypescriptOptions
) {
  const {
    bannerComment = defaultBannerComment,
    format = true,
    unreachableDefinitions = false,
    strictIndexSignatures = true,
  } = options || {};

  const json = objectToJSON(name, object);

  return jsonToTypescript(json, name, {
    bannerComment,
    format,
    unreachableDefinitions,
    strictIndexSignatures,
  });
}
