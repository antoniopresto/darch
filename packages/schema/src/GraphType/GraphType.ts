import { RuntimeError } from '@darch/utils/lib/RuntimeError';
import { StrictMap } from '@darch/utils/lib/StrictMap';
import { assertSame } from '@darch/utils/lib/assertSame';
import { isProduction } from '@darch/utils/lib/env';
import type {
  GraphQLFieldConfig,
  GraphQLInterfaceType,
  GraphQLNamedInputType,
  GraphQLNamedType,
  GraphQLResolveInfo,
} from 'graphql';

import { Infer } from '../Infer';
import { createObjectType, ObjectType } from '../ObjectType';
import { FieldDefinitionConfig } from '../TObjectConfig';
import { TAnyFieldType } from '../fields/FieldType';
import { GraphTypeLike } from '../fields/IObjectLike';
import { getObjectDefinitionId } from '../fields/MetaFieldField';
import { ObjectField } from '../fields/ObjectField';
import {
  ObjectDefinitionInput,
  ObjectFieldInput,
  ToFinalField,
} from '../fields/_parseFields';
import type { ObjectToTypescriptOptions } from '../objectToTypescript';
import { parseObjectField } from '../parseObjectDefinition';

import type { ConvertFieldResult, GraphQLParserResult } from './GraphQLParser';

const register = new StrictMap<string, any>();
const resolvers = new StrictMap<string, any>();

export class GraphType<Definition> {
  static __isGraphType = true;
  __isGraphType = true;

  static register = register;
  static resolvers = resolvers;

  static reset = async () => {
    resolvers.clear();
    register.clear();
  };

  readonly definition: ToFinalField<Definition>;

  __field: TAnyFieldType;

  readonly id: string;
  readonly _object?: ObjectType<any>;

  constructor(
    definition: Definition extends ObjectFieldInput ? Definition : never
  );

  constructor(
    name: string,
    definition: Definition extends ObjectFieldInput ? Definition : never
  );

  constructor(...args: any[]) {
    let name: string | undefined = undefined;
    let definition: ObjectFieldInput;

    if (args.length === 2) {
      name = args[0];
      definition = args[1];
    } else {
      definition = args[0];
    }

    this.__field = parseObjectField('temp', definition, true);

    if (ObjectField.is(this.__field)) {
      if (
        name &&
        this.__field.utils.object.id &&
        this.__field.utils.object.id !== name
      ) {
        this.__field.utils.object = this.__field.utils.object.clone(name);
      } else if (name) {
        this.__field.utils.object.identify(name);
      } else {
        name = getObjectDefinitionId(this.__field.utils.object.definition);
      }

      this._object = this.__field.utils.object;
    }

    if (!name) {
      throw new RuntimeError(`Expected name to be provided, found ${name}`, {
        name,
        definition,
      });
    }

    this.id = name;

    this.definition = this.__field.toObjectFieldType() as any;

    if (register.has(name)) {
      const existing = register.get(name);

      if (!isProduction()) {
        assertSame(
          `Different type already registered with name "${name}"`,
          this.definition,
          existing
        );
      }
    } else {
      register.set(name, this.definition as any);
    }
  }

  parse = (
    ...args: Parameters<TAnyFieldType['parse']>
  ): Infer<ToFinalField<Definition>> => {
    return this.__field.parse(...args);
  };

  _toGraphQL = () => {
    return ObjectType.serverUtils().graphqlParser.GraphQLParser.fieldToGraphQL({
      field: this.__field,
      path: [`Type_${this.id}`],
      plainField: this.__field.toObjectFieldType(),
      fieldName: this.id,
      parentName: this.id,
    });
  };

  graphQLType = (
    ...args: Parameters<ConvertFieldResult['type']>
  ): GraphQLNamedType => {
    return this._toGraphQL().type(...args) as any;
  };

  graphQLInputType = (
    ...args: Parameters<ConvertFieldResult['inputType']>
  ): GraphQLNamedInputType => {
    return this._toGraphQL().inputType(...args) as any;
  };

  graphQLInterface = (
    ...args: Parameters<GraphQLParserResult['interfaceType']>
  ): GraphQLInterfaceType => {
    if (!this._object) {
      throw new RuntimeError(
        'graphQLInterface is only available for object type',
        {
          type: this.__field.type,
        }
      );
    }

    return ObjectType.serverUtils()
      .graphqlParser.GraphQLParser.objectToGraphQL({
        object: this._object,
      })
      .interfaceType(...args);
  };

  relations = {};

  addRelation = <
    FieldTypeDef extends ObjectFieldInput,
    Name extends string,
    Context = unknown,
    ArgsDef extends ObjectDefinitionInput = ObjectDefinitionInput
  >(
    options: DarchGraphQLFieldConfigInput<
      Context,
      Infer<ToFinalField<Definition>>,
      FieldTypeDef,
      ArgsDef
    > & { type: FieldTypeDef; name: Name }
  ): this => {
    const object = this._object;

    const type = createType(options.name, options.type);

    const { name } = options;

    if (!object) {
      throw new RuntimeError(`Can't add relation to a not object type`, {
        object,
        type,
        options,
      });
    }

    const resolver = type.createResolver({
      ...(options as any),
    });

    resolver.__isRelation = true;
    resolver.__relatedToGraphTypeId = this.id;

    this._object?.addGraphQLMiddleware((hooks) => {
      hooks.onFieldConfigMap.register(function (fields) {
        fields[name] = resolver;
      });
    });

    return this;
  };

  print = (): string[] => {
    const type = this.graphQLType();
    const inputType = this.graphQLInputType();

    const { GraphQLSchema, printSchema } = ObjectType.serverUtils().graphql;

    const object = new GraphQLSchema({
      // @ts-ignore
      types: [type, inputType],
    });

    return printSchema(object).split('\n');
  };

  typescriptPrint = (
    options?: ObjectToTypescriptOptions & { name?: string }
  ) => {
    const object =
      this._object ||
      createObjectType({
        [this.id]: this.definition,
      });

    return ObjectType.serverUtils().objectToTypescript.objectToTypescript(
      options?.name || this.id,
      object,
      options
    );
  };

  createResolver = <
    Context = unknown,
    Source = unknown,
    ArgsDef extends
      | ObjectDefinitionInput
      | Readonly<ObjectDefinitionInput> = ObjectDefinitionInput
  >(
    options: DarchGraphQLFieldConfigInput<Context, Source, Definition, ArgsDef>
  ): DarchResolver<Context, Source, Definition, ArgsDef> => {
    const { args, name = this.id, kind = 'query', resolve, ...rest } = options;

    if (resolvers.has(name)) {
      return resolvers.get(name);
    }

    const parsePayload = this.parse.bind(this);

    const argsObject = isPossibleArgsDef(args)
      ? createObjectType(`${name}Input`, args)
      : undefined;

    const ArgsType: any = argsObject
      ? argsObject.graphqlInputType({ name: `${name}Input` })
      : undefined;

    const resolveFunction: any = async function typeCheckResolveWrapper(
      source,
      args: any,
      context: any,
      info: any
    ): Promise<any> {
      args = argsObject
        ? argsObject.parse(args, {
            customMessage: (_, error) => {
              return `Invalid input provided to resolver "${name}":\n ${error.message}`;
            },
          })
        : args;

      const result = await resolve(source, args, context, info);

      return parsePayload(
        result,
        (_, error) => `Invalid output from resolver "${name}": ${error.message}`
      );
    };

    const type = this.graphQLType();

    const result: any = {
      ...rest,
      kind,
      name,
      resolve: resolveFunction,
      args: ArgsType?.getFields() || {},
      type,
      argsDef: args,
      typeDef: this.definition,
      __isResolver: true,
      __isRelation: false,
      __relatedToGraphTypeId: '',
      __graphTypeId: this.id,
    };

    resolvers.set(name, result);

    return result as any;
  };

  /**
   * Get an Object with the provided id
   *    or set a new Object in the register if not found.
   * @param id
   * @param def
   */
  static getOrSet = <T extends FieldDefinitionConfig>(
    id: string,
    def: T
  ): GraphType<T> => {
    const existing =
      GraphType.register.has(id) &&
      (GraphType.register.get(id) as GraphType<T>);

    if (existing) return existing;

    return new GraphType<any>(id, def) as any;
  };

  static is(input: any): input is GraphType<any> {
    return input?.__isGraphType === true;
  }

  static isTypeDefinition(
    input: any
  ): input is { type: GraphTypeLike; list?: boolean; optional?: boolean } {
    return input?.type?.__isGraphType === true;
  }
}

export function createType<Definition extends ObjectFieldInput>(
  definition: Definition
): GraphType<Definition>;

export function createType<Definition extends ObjectFieldInput>(
  name: string,
  definition: Definition
): GraphType<Definition>;

export function createType(...args: any[]) {
  return new GraphType(
    // @ts-ignore
    ...args
  );
}

export interface DarchGraphQLFieldConfigInput<
  Context = unknown,
  Source = unknown,
  TypeDef = unknown,
  ArgsDef = unknown
> extends Omit<
    GraphQLFieldConfig<Source, Context>,
    'resolve' | 'type' | 'args'
  > {
  name: string;
  kind?: 'query' | 'mutation' | 'subscription';
  args?: ArgsDef;
  resolve: ResolveFunction<Context, Source, TypeDef, ArgsDef>;
}

export type AnyDarchResolver = DarchResolver;

export interface DarchResolver<
  Context = unknown,
  Source = unknown,
  TypeDef = unknown,
  ArgsDef = unknown
> extends Omit<GraphQLFieldConfig<Source, Context>, 'resolve'> {
  __isResolver: true;
  __isRelation: boolean;
  __graphTypeId: string;
  __relatedToGraphTypeId: string;

  resolve: ResolveFunction<Context, Source, TypeDef, ArgsDef>;
  name: string;
  kind: 'query' | 'subscription' | 'mutation';
  typeDef: TypeDef extends ObjectFieldInput ? TypeDef : never;
  argsDef: ArgsDef extends Record<string, any> ? ArgsDef : never;
}

export interface ResolveFunction<
  Context = unknown,
  Source = unknown,
  TypeDef = unknown,
  ArgsDef = unknown
> {
  (
    source: Source,
    args: ArgsDef extends { [K: string]: ObjectFieldInput }
      ? Infer<ArgsDef>
      : {},
    context: Context,
    info: GraphQLResolveInfo
  ): Promise<Infer<ToFinalField<TypeDef>>>;
}

function isPossibleArgsDef(args: any): args is Readonly<ObjectDefinitionInput> {
  return args && typeof args === 'object' && Object.keys(args).length;
}
