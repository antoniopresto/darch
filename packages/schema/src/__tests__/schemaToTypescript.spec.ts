import { createType } from '../GraphType/GraphType';
import { createObjectType } from '../ObjectType';
import {
  createGraphQLSchema,
  resolversTypescriptParts,
} from '../createGraphQLSchema';
import { objectToTypescript } from '../objectToTypescript';

describe('objectToTypescript', () => {
  it('works', async () => {
    const object = createObjectType({
      name: 'string',
      date: 'date',
      pointers: '[cursor]',
      bool: 'boolean?',
      ulids: '[ulid]?',
      dates: '[date]',
      id: '[ID]?',
      sex: { union: [{ enum: ['m', 'f', 'o'], optional: true }] },
      addresses: {
        union: [
          {
            object: {
              kind: { enum: ['home'] },
              street: 'string',
              number: { union: ['string', 'float'] },
            },
            optional: true,
            list: true,
            description: 'Home address',
          },
          {
            object: {
              kind: { enum: ['work'] },
              week_days: { enum: ['0', '1', '2', '3', '4', '5', '6'] },
              name: {
                object: {
                  firstName: 'string',
                  lastName: 'string',
                },
              },
            },
          },
        ],
      },
    } as const)
      .describe('My Custom Object')
      .describe({ name: 'person name', pointers: 'some pointers' });

    const ts = await objectToTypescript('MyObject', object);

    expect(ts.split('\n')).toEqual([
      '/**',
      ' * My Custom Object',
      ' */',
      'export interface MyObject {',
      '  /**',
      '   * person name',
      '   */',
      '  name: string;',
      '  date: Date;',
      '  pointers: Cursor[];',
      '  bool?: boolean;',
      '  ulids?: Ulid[];',
      '  dates: Date[];',
      '  id?: ID[];',
      '  sex?: "m" | "f" | "o";',
      '  addresses?:',
      '    | {',
      '        kind: "home";',
      '        street: string;',
      '        number: string | number;',
      '      }[]',
      '    | {',
      '        kind: "work";',
      '        week_days: "0" | "1" | "2" | "3" | "4" | "5" | "6";',
      '        name: {',
      '          firstName: string;',
      '          lastName: string;',
      '        };',
      '      };',
      '}',
      '',
    ]);
  });

  it('handle union', async () => {
    const IntervalType = createType('Interval', {
      union: [
        // (10|7|5|3)
        {
          object: {
            from: { literal: '10' },
          },
        },
        {
          object: {
            from: { literal: '7' },
          },
        },
        {
          object: {
            from: { literal: '5' },
          },
        },
        {
          object: {
            from: { literal: '3' },
          },
        },
      ],
    });

    const resolver = IntervalType.createResolver({
      name: 'getIntervals',
      async resolve() {
        return IntervalType.parse({});
      },
    });

    const schema = createGraphQLSchema([resolver]);
    const ts = await resolversTypescriptParts({
      name: 'Tools',
      options: {},
      resolvers: schema.utils.resolvers,
    });

    expect(ts.code.split('\n')).toEqual([
      'export type getIntervalsInput = undefined;',
      'export type Interval =',
      '  | {',
      '      from: 10;',
      '    }',
      '  | {',
      '      from: 7;',
      '    }',
      '  | {',
      '      from: 5;',
      '    }',
      '  | {',
      '      from: 3;',
      '    };',
      'export interface Tools {',
      '  getIntervals: { input: getIntervalsInput; payload: Interval };',
      '}',
      'export type QueryResolvers = {',
      '  getIntervals(args: getIntervalsInput): Promise<Interval>;',
      '};',
      'export type MutationResolvers = {};',
      'export type SubscriptionResolvers = {};',
      '',
    ]);
  });
});
