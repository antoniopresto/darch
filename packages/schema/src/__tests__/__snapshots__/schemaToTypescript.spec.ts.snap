// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaToTypescript works 1`] = `
"/* tslint:disable */
/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

/**
 * My Custom Schema
 */
export interface MySchema {
  /**
   * person name
   */
  name: string;
  date: Date;
  pointers: Cursor[];
  bool?: boolean;
  ulids?: Ulid[];
  dates: Date[];
  sex: \\"m\\" | \\"f\\" | \\"o\\";
  addresses?:
    | {
        kind: \\"home\\";
        street: string;
        number: string | number;
      }[]
    | {
        kind: \\"work\\";
        week_days: \\"0\\" | \\"1\\" | \\"2\\" | \\"3\\" | \\"4\\" | \\"5\\" | \\"6\\";
        name: {
          firstName: string;
          lastName: string;
        };
      };
}
"
`;
